crates/miden-lib/src/note/well_known_note.rs

```diff
@@ -157,4 +163,242 @@ impl WellKnownNote {
+    ///       either consumed, or consumed after timelock height, or consumed after reclaim height.
+    ///       See the underlying table for the specific cases. Notice that this table defines all
+    ///       possible variations of the values of the current block height, timelock height and
+    ///       reclaim height relative to each other (it was greatly reduced though).
+    /// ```text
+    ///       Where:
+    ///         - `curr` -- current block height
+    ///         - `tl`   -- timelock height
+    ///         - `rc`   -- reclaim height
+    ///         - `v`    -- symbol to show that heights have any relation
+    ///       ┏━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━┓
+    ///       ┃ # ┃  Height sequence   ┃   Sender    ┃  Receiver   ┃
+    ///       ┣━━━╋━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━╋━━━━━━━━━━━━━┫
+    ///       │ 1 │  rc  v  tl  ≤ curr │ return None | return None |
+    ///       ├───┼────────────────────┼─────────────┼─────────────┤
+    ///       │ 2 │ curr v  rc  <  tl  │  return tl  |  return tl  |
+    ///       ├───┼────────────────────┼─────────────┼─────────────┤
+    ///       │ 3 │ curr <  tl  ≤  rc  │  return rc  |  return tl  |
+    ///       ├───┼────────────────────┼─────────────┼─────────────┤
+    ///       | 4 │  tl  ≤ curr <  rc  │  return rc  | return None |
+    ///       └───┴────────────────────┴─────────────┴─────────────┘
+    /// ```
+    fn check_note_inputs_inner(
+        &self,
+        note: &Note,
+        target_account_id: AccountId,
+        block_ref: BlockNumber,
+    ) -> Result<Option<NoteConsumptionStatus>, StaticAnalysisError> {
+        match self {
+            WellKnownNote::P2ID => {
+                let note_inputs = note.inputs().values();
+                if note_inputs.len() != self.num_expected_inputs() {
+                    return Err(StaticAnalysisError::new(format!(
+                        "P2ID note should have {} inputs, but {} was provided",
+                        WellKnownNote::P2ID.num_expected_inputs(),
+                        note_inputs.len()
+                    )));
+                }
+
+                let input_account_id = try_read_account_id_from_inputs(note_inputs)?;
+
+                if input_account_id == target_account_id {
+                    Ok(None)
+                } else {
+                    Ok(Some(NoteConsumptionStatus::NeverConsumable("account ID provided to the P2ID note inputs doesn't match the target account ID".into())))
+                }
+            },
+            WellKnownNote::P2IDE => {
+                let note_inputs = note.inputs().values();
+                if note_inputs.len() != self.num_expected_inputs() {
+                    return Err(StaticAnalysisError::new(format!(
+                        "P2IDE note should have {} inputs, but {} was provided",
+                        WellKnownNote::P2IDE.num_expected_inputs(),
+                        note_inputs.len()
+                    )));
+                }
```

Comment:
Could we encapsulate the logic for input parsing in helper functions? For example, something like:
```rust
fn parse_p2id_inputs(inputs: &NoteInputs) -> Result<AccountId, SomeError> {
    ...
}

fn parse_p2ide_inputs(inputs: &NoteInputs) -> Rsult<(AccountId, u32, u32), SomeError> {
    ...
}
```

It would also be good to re-organize the `well_known_note` module into a directory that looks something like:
```
note
└── well_known_notes
    ├── mod.rs
    ├── p2id.rs
    ├── p2ide.rs
    ├── swap.rs
    └── etc.
```
But maybe that's for another PR.